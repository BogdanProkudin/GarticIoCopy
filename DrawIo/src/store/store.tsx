import { configureStore } from "@reduxjs/toolkit";
import userAuth from "./slices/userAuth";
import drawThema from "./slices/roomInfo";
import userInfo from "./slices/userInfo";
import drawInfo from "./slices/drawInfo";
export const store = configureStore({
  reducer: {
    userAuth: userAuth,
    drawThema: drawThema,
    userInfo: userInfo,
    drawInfo: drawInfo,
  },
});
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
