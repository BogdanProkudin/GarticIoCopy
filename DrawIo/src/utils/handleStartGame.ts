import axios from "axios";
import { socket } from "../socket";
import { IRoomUsers, setIsIntervalOver } from "../store/slices/roomInfo";
interface IhandleGameStartButton {
  roomUsers: IRoomUsers[];
  activeIndex: number;
  roomId: string;
  words: string[];
  dispatch: any;
  maxGamePoints: number;
}
export const handleGameStartButton = async ({
  roomId,
  roomUsers,
  activeIndex,
  maxGamePoints,
  words,
  dispatch,
}: IhandleGameStartButton) => {
  const sendGameStart = async (url: string) => {
    try {
      await axios.post(url, {
        headers: {
          "Content-Type": "application/json",
        },
        users: roomUsers,
        roomId,
        maxGamePoints,
      });
      // socket.emit("usersNotGuessedTimer", roomId);
    } catch (error) {
      console.error("Error:", error);
    }
  };
  const sendYourTurn = async (url: string) => {
    try {
      const response = await axios.post(url, {
        headers: {
          "Content-Type": "application/json",
        },
        roomId,
        activeIndex,
        roomUsers,
      });
      dispatch(setIsIntervalOver(response.data.timer));
    } catch (error) {
      console.error("Error:", error);
    }
  };

  const url = "http://localhost:3000/startNewGame";

  await sendGameStart(url);

  socket.emit("startGame", { words: words, roomId });
  const url2 = "http://localhost:3000/yourTurn";
  await sendYourTurn(url2);
};
