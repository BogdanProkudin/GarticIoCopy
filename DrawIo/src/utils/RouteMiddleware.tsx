import { useState, useEffect } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import { useGetRoomIdFromUrl } from "../hooks/useGetRoomIdFromUrl";
import { socket } from "../socket";
import LottieSettings from "../tools/Animation - 1725791635271.json";
import styles from "../components/setUpPage/styles.module.scss";
import { useAppDispatch, useAppSelector } from "../store/hook";
import {
  getRoomData,
  setRoomData,
  setSelectedThema,
  setRoomUsers,
  setSelectedPlayers,
  setHost,
  setMaxRoomPoints,
  IRoomData,
  setIsGameRoomLoading,
} from "../store/slices/roomInfo";
import Lottie from "react-lottie";

const RouteMiddleware = ({ children }: any) => {
  const navigate = useNavigate();
  const isGameRoomLoading = useAppSelector(
    (state) => state.drawThema.isGameRoomLoading
  );
  const location = useLocation();
  const userId = localStorage.getItem("userId");

  const [isChecking, setIsChecking] = useState(true); // Флаг проверки
  const [isError, setIsError] = useState(false);

  const dispatch = useAppDispatch();
  const defaultOptions = {
    loop: true,
    autoplay: true,
    animationData: LottieSettings,
    rendererSettings: {
      preserveAspectRatio: "xMidYMid slice",
    },
  };

  const roomId = useGetRoomIdFromUrl();
  const userNameStorage = localStorage.getItem("userName");

  const getUserLeave = async () => {
    try {
      const isGamePage = location.pathname === `/game/${roomId}`;
      if (isGamePage) {
        console.log("IS LOAD", isGameRoomLoading);

        dispatch(setIsGameRoomLoading(true)); // Ставим флаг загрузки

        const response = await dispatch(getRoomData({ roomId, userId }));

        if (response?.payload.message === "Room does not exist") {
          setIsChecking(false); // Завершаем проверку
          return "lobbyNotFound";
        }

        const roomDataFromResponse = response?.payload?.[0];
        if (!roomDataFromResponse) {
          setIsChecking(false); // Завершаем проверку
          return "lobbyNotFound";
        }

        const currentUser = roomDataFromResponse.usersInfo.find(
          (user: any) => user.userId === userId
        );

        // Если пользователь — гость, перенаправляем его на prepareRoom
        if (
          !roomDataFromResponse.usersInfo.some(
            (user: any) => user.userId === userId
          ) &&
          userId !== roomDataFromResponse.host.hostId
        ) {
          dispatch(setRoomData(roomDataFromResponse));
          setIsChecking(false); // Завершаем проверку
          return "prepare";
        }

        // Обновляем данные комнаты
        dispatch(setRoomData(roomDataFromResponse));
        dispatch(setSelectedThema(roomDataFromResponse.thema));
        dispatch(setRoomUsers(roomDataFromResponse.usersInfo));
        dispatch(setSelectedPlayers(roomDataFromResponse.players));
        dispatch(setHost(roomDataFromResponse.host));
        dispatch(setMaxRoomPoints(roomDataFromResponse.points));

        if (currentUser?.isUserInLobby) {
          setIsChecking(false); // Завершаем проверку
          return "prepare";
        }

        if (currentUser?.isUserLeave) {
          setIsChecking(false); // Завершаем проверку
          return "main";
        }

        localStorage.setItem("pageAccessedByReload", "true");
      } else {
        if (location.pathname === `/prepareRoom/${roomId}`) {
          setIsChecking(false); // Завершаем проверку
        }
        return "game";
      }
    } catch (err) {
      console.log("ERROR NAVIGATION", err);
      setIsError(true);
      return "lobbyNotFound";
    } finally {
      // Завершаем проверку
      setIsChecking(false);
      dispatch(setIsGameRoomLoading(false));
    }
  };

  useEffect(() => {
    const handleUserJoined = (data: IRoomData) => {
      dispatch(setRoomUsers(data.usersInfo));
      console.log("User joined", data);
    };

    socket.on("userJoined", handleUserJoined);

    const func = async () => {
      const routeOutcome = await getUserLeave();
      if (routeOutcome === "main") {
        navigate("/");
      }
      if (routeOutcome === "prepare") {
        navigate(`/prepareRoom/${roomId}`, { replace: true });
      }
      if (routeOutcome === "lobbyNotFound") {
        navigate(`/lobbyNotFound`, { replace: true });
      }
    };
    func();

    return () => {
      socket.off("userJoined", handleUserJoined);
    };
  }, [location, navigate]);

  // Если проверка еще выполняется, ничего не рендерим
  if (isChecking || isGameRoomLoading) {
    return (
      <div className={styles.set_up_loading}>
        <Lottie
          isClickToPauseDisabled
          style={{ height: "300px", width: "300px" }}
          options={defaultOptions}
        />
      </div>
    );
  }

  // Обработка ошибок
  if (isError) {
    return <div className={styles.error_message}>Room not found</div>;
  }

  return children;
};

export default RouteMiddleware;
