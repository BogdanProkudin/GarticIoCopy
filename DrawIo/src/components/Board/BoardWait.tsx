import Lottie from "react-lottie";
import { useAppDispatch, useAppSelector } from "../../store/hook";
import styles from "./styles.module.scss";
import LottieSettings from "../../tools/Animation - 1711477541401.json";
import { socket } from "../../socket";
import { setIsGameStarted, setIsRoundEnd } from "../../store/slices/roomInfo";
import { memo, useEffect } from "react";
import { handleGameStartButton } from "../../utils/handleStartGame";
import { setIsUserDraw } from "../../store/slices/userInfo";
import { useGetRoomIdFromUrl } from "../../hooks/useGetRoomIdFromUrl";

export const useGameSelectors = () => {
  return {
    activeIndex: useAppSelector((state) => state.drawThema.activeIndex),
    isGameStarted: useAppSelector((state) => state.drawThema.isGameStarted),
    roomUsers: useAppSelector((state) => state.drawThema.roomUsers),
    host: useAppSelector((state) => state.drawThema.host),
    selectedThema: useAppSelector((state) => state.drawThema.selectedThema),
  };
};

const BoardWait = () => {
  const dispatch = useAppDispatch();
  const roomId = useGetRoomIdFromUrl();
  const { activeIndex, roomUsers, host, selectedThema } = useGameSelectors();
  const userNameStorage = localStorage.getItem("userName");

  const maxGamePoints = useAppSelector(
    (state) => state.drawThema.maxRoomPoints
  );
  const defaultOptions = {
    loop: true,
    autoplay: true,
    animationData: LottieSettings,
    rendererSettings: {
      preserveAspectRatio: "xMidYMid slice",
    },
  };
  if (!roomUsers) {
    return <div>IN BOARD WAIT ZERO USERS</div>;
  }

  useEffect(() => {
    const onGameStarted = () => {
      dispatch(setIsUserDraw(true));
      dispatch(setIsRoundEnd(true));

      dispatch(setIsGameStarted(true));
    };

    socket.on("gameStarted", onGameStarted);

    return () => {
      socket.off("gameStarted", onGameStarted);
    };
  }, [dispatch]);

  const waitingTextStyles = {
    top:
      roomUsers.length > 1 && host.hostName !== userNameStorage ? "3rem" : "",
  };

  const animationContainerStyles = {
    top:
      roomUsers.length > 1 && host.hostName !== userNameStorage
        ? "5rem"
        : "2rem",
  };

  const waitingForOwnerTextStyles = {
    top:
      roomUsers.length > 1 && host.hostName !== userNameStorage
        ? "21.5rem"
        : "",
  };

  return (
    <div className={styles.board_wait_container}>
      <h1 className={styles.game_room_wait_text} style={waitingTextStyles}>
        WAITING
      </h1>
      <div
        className={styles.board_wait_animation}
        style={animationContainerStyles}
      >
        <Lottie
          options={defaultOptions}
          height={190}
          width={280}
          style={{ marginTop: "3.5rem" }}
          isClickToPauseDisabled
        />
      </div>
      {roomUsers.length <= 1 && (
        <span className={styles.game_room_wait_small_text}>
          Waiting for players
        </span>
      )}
      {roomUsers.length > 1 && host.hostName === userNameStorage && (
        <>
          <span className={styles.game_room_wait_small_text}>
            Players are waiting for you to start
          </span>
          <button
            data-testid={"button-id"}
            onClick={() =>
              handleGameStartButton({
                words: selectedThema.words,
                roomUsers,
                maxGamePoints,
                roomId,
                activeIndex,
                dispatch: dispatch,
              })
            }
            className={styles.game_room_start_button}
          >
            <div />
            <strong>PLAY!</strong>
          </button>
        </>
      )}
      {roomUsers.length > 1 && host.hostName !== userNameStorage && (
        <span
          className={styles.game_room_wait_small_text}
          style={waitingForOwnerTextStyles}
        >
          Waiting for the owner to start
        </span>
      )}
    </div>
  );
};

export default memo(BoardWait);
