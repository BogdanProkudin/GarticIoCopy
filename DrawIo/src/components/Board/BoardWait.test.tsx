import BoardWait from "./BoardWait";

import { render, screen, fireEvent } from "@testing-library/react";
import { useAppSelector } from "../../store/hook";
// import { testUseBoardWaitSelector } from "./test-boardWait-selector";
import { handleGameStartButton } from "../../utils/handleStartGame";

jest.mock("react-redux");
jest.mock("../../utils/handleStartGame", () => ({
  handleGameStartButton: jest.fn(),
}));
describe("Board Wait test", () => {
  const state = {
    drawThema: {
      roomUsers: [
        { id: "1", userName: "owner", isActive: false },
        { id: "2", userName: "owner2", isActive: false },
      ],
      host: "owner",
      activeIndex: 0,
      selectedThema: {
        name: "Животные",
        icon: "",
        words: ["собака", "крокодил"],
      },
    },
  };
  const stateForAloneInRoom = {
    drawThema: {
      roomUsers: [{ id: "1", userName: "owner", isActive: false }],
      host: "owner",
      activeIndex: 0,
      selectedThema: {
        name: "Животные",
        icon: "",
        words: ["собака", "крокодил"],
      },
    },
  };

  const mockUseAppSelector = useAppSelector as jest.MockedFunction<
    typeof useAppSelector
  >;

  beforeEach(() => {});

  it("if host name === userNameStorageName ", () => {
    mockUseAppSelector.mockImplementation((f: any) => f(state));
    Storage.prototype.getItem = jest.fn((key) => {
      return key === "userName" ? "owner" : null;
    });
    const { getByTestId } = render(<BoardWait />);

    const linkElement = screen.getByText(
      "Players are waiting for you to start"
    );

    expect(linkElement).toBeInTheDocument;
    const button = getByTestId("button-id");
    fireEvent.click(button);

    // Проверяем, что dispatch был вызван
    expect(handleGameStartButton).toHaveBeenCalled();
  });
  it("if host name !== userNameStorageName", () => {
    Storage.prototype.getItem = jest.fn((key) => {
      return key === "userName" ? " not owner" : null;
    });
    render(<BoardWait />);
    const linkElement = screen.getByText("Waiting for the owner to start");

    expect(linkElement).toBeInTheDocument;
  });
  it("if user alone in room", () => {
    mockUseAppSelector.mockImplementation((f: any) => f(stateForAloneInRoom));
    render(<BoardWait />);
    const linkElement = screen.getByText("Waiting for players");

    expect(linkElement).toBeInTheDocument;
  });
});
