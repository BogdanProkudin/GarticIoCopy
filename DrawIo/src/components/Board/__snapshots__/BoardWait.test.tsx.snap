// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Wait TEst renders if good 1`] = `
<span
  style="top: 21.5rem;"
>
  Waiting for the owner to start
</span>
`;

exports[`should render board with 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          WAITING
        </h1>
        <div
          style="top: 2rem;"
        >
          <div
            aria-label="animation"
            role="button"
            style="width: 100%; height: 100%; overflow: hidden; margin: 0px auto; outline: none;"
            tabindex="0"
            title=""
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        WAITING
      </h1>
      <div
        style="top: 2rem;"
      >
        <div
          aria-label="animation"
          role="button"
          style="width: 100%; height: 100%; overflow: hidden; margin: 0px auto; outline: none;"
          tabindex="0"
          title=""
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
